engine/src/juliabox/plugins/compute_gce/impl_gce.py

```diff
@@ -0,0 +1,633 @@
+
+    @staticmethod
+    def _connect_google_monitoring():
+        if CompGCE.MONITORING_CONN is None:
+            CompGCE.MONITORING_CONN = build("cloudmonitoring", "v2beta2",
+                                            credentials=CompGCE._get_service_credentials())
+        return CompGCE.MONITORING_CONN
+
+    @staticmethod
+    @retry_on_bsl
+    def _instance_attrs(instance_name=None):
+        if instance_name is None:
+            instance_name = CompGCE.get_instance_id()
+        ins = CompGCE._connect_gce().instances()
+        return ins.get(project=CompGCE.INSTALL_ID, zone=CompGCE._zone(),
+                       instance=instance_name).execute()
+
+    @staticmethod
+    def _uptime_minutes(instance_name=None):
+        attrs = CompGCE._instance_attrs(instance_name)
+        lt = parse_iso_time(attrs["creationTimestamp"])
+        nt = datetime.datetime.now(pytz.utc)
+        uptime = nt - lt
+        minutes = int(uptime.total_seconds()/60)
+        return minutes
+
+    @staticmethod
+    @retry_on_bsl
+    def _get_instances(gname, only_running=False):
+        conn = CompGCE._connect_gce().instanceGroups()
+        flag = 'RUNNING' if only_running else 'ALL'
+        data = conn.listInstances(project=CompGCE.INSTALL_ID, zone=CompGCE._zone(),
+                                  body={'instanceState': flag},
+                                  instanceGroup=gname).execute()
+        ret = []
+        for inst in data.get('items', []):
+            ret.append(inst['instance'].split('/')[-1])
+        return ret
+
+    @staticmethod
+    def get_all_instances(gname=None):
+        if gname is None:
+            gname = CompGCE.AUTOSCALE_GROUP
+        if gname is None:
+            return [CompGCE.get_instance_id()]
+        instances = CompGCE._get_instances(gname, only_running=True)
+        if len(instances) == 0:
+            return [CompGCE.get_instance_id()]
+        return instances
+
+    @staticmethod
+    @retry_on_bsl
+    def _increment_num_instances():
+        conn = CompGCE._connect_gce().instanceGroupManagers()
+        curr = conn.get(project=CompGCE.INSTALL_ID, zone=CompGCE._zone(),
+                        instanceGroupManager=CompGCE.AUTOSCALE_GROUP).execute()['targetSize']
+        return conn.resize(project=CompGCE.INSTALL_ID, zone=CompGCE._zone(),
+                           instanceGroupManager=CompGCE.AUTOSCALE_GROUP,
+                           size=curr + 1).execute()
+
+    @staticmethod
+    def _add_instance():
+        try:
+            # Execute policy only after a reasonable wait period to let a new machine boot up.
+            # This will prevent thrashing AWS APIs and triggering AWS throttling.
+            # Cooldown policy can also apply after that.
+            now = datetime.datetime.utcnow()
+            if (CompGCE.LAST_SCALE_UP_TIME is None) or \
+               (now > CompGCE.LAST_SCALE_UP_TIME + datetime.timedelta(minutes=5)):
+                CompGCE.LAST_SCALE_UP_TIME = now
```

Comment:
`LAST_SCALE_UP_TIME` is local to the process. As this is not shared across JuliaBox instances, each existing instance can potentially bring up a new instance when loaded.

This may not be the intended behavior.